src/OpenFeature/OpenFeatureClient.cs

```diff
@@ -0,0 +1,150 @@
+                defaultValue,
+                FlagValueType.Boolean,
+                _metadata,
+                OpenFeature.GetProviderMetadata(),
+                context
+            );*/
+
+            FlagEvaluationDetails<T> response;
+            try
+            {
+                //TriggerHook(HookLifeCycle.Before, allHooks, hookContext, config);
+
+                // TODO Extract this out into another method so this method isn't too complex
+                switch (flagValueType)
+                {
+                    case FlagValueType.Boolean:
+                        response = provider.ResolveBooleanValue(flagKey, (bool)Convert.ChangeType(defaultValue, typeof(bool)),
+                            context, options).ToFlagEvaluationDetails() as FlagEvaluationDetails<T>;
+                        break;
+                    case FlagValueType.Number:
+                        response = provider.ResolveNumberValue(flagKey, (int)Convert.ChangeType(defaultValue, typeof(int)),
+                            context, options).ToFlagEvaluationDetails() as FlagEvaluationDetails<T>;
+                        break;
+                    case FlagValueType.String:
+                        response = provider.ResolveStringValue(flagKey, (string)Convert.ChangeType(defaultValue, typeof(string)),
+                            context, options).ToFlagEvaluationDetails() as FlagEvaluationDetails<T>;
+                        break;
+                    case FlagValueType.Object:
+                        response = provider.ResolveStructureValue(flagKey, defaultValue, context, options).ToFlagEvaluationDetails();
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException(nameof(flagValueType), flagValueType, "Unknown flag value type");
+                }
+
+
+                // hook after
+                //TriggerHook(HookLifeCycle.After, allHooksReversed, hookContext, config, null, response);
+            }
+            catch (Exception e)
+            {
+                //Error hook callback
+                //TriggerHook(HookLifeCycle.Error, allHooksReversed, hookContext, config, e);
+                response = new FlagEvaluationDetails<T>(flagKey, defaultValue, e.Message, Reason.Error, string.Empty);
+            }
+            finally
+            {
+                //TriggerHook(HookLifeCycle.Finally, allHooksReversed, hookContext, config);
+            }
+
+            return response;
+        }
+
+        private static void TriggerHook<T>(HookLifeCycle lifeCycle, IEnumerable<IHook> hooks, HookContext<T> context,
+            FlagEvaluationOptions options, Exception error = null, FlagEvaluationDetails<T> evaluationDetails = null)
+        {
+            foreach (var hook in hooks)
+            {
+                switch (lifeCycle)
+                {
+                    case HookLifeCycle.Before:
+                        hook.Before(context, options.HookHints);

```

Comment:
Yea I will need to async the provider interface as well. We will probably need to provide both sync and async versions of the API. What are your thoughts on this? Standard for c# is to prefix any async methods with `{METHOD_NAME}Async`

Reply:
> Thinking about it I don't think there is much point in supporting synchronous, async model was introduced in 2012 in c# language version 5. This is the defacto standard now. I'll address the async under hooks implementation PR instead of this.

I think this is the right direction to go.